src/metrics.cpp

```diff
@@ -203,16 +203,56 @@ void ConnectMetricsScreen()
     uiInterface.InitMessage.connect(metrics_InitMessage);
 }
 
+std::string DisplayTime(int64_t time, TimeFormat format)
+{
+    int days =  time / (24 * 60 * 60);
+    int hours = (time - (days * 24 * 60 * 60)) / (60 * 60);
+    int minutes = (time - (((days * 24) + hours) * 60 * 60)) / 60;
+    int seconds = time - (((((days * 24) + hours) * 60) + minutes) * 60);
+
+    std::string strTime;
+    if (format == TimeFormat::REDUCED) {
+        if (days > 0) {
+            strTime = strprintf(_("%d days"), days);
+        } else if (hours > 0) {
+            strTime = strprintf(_("%d hours"), hours);
+        } else if (minutes > 0) {
+            strTime = strprintf(_("%d minutes"), minutes);
+        } else {
+            strTime = strprintf(_("%d seconds"), seconds);
+        }
+    } else {
+        if (days > 0) {
+            strTime = strprintf(_("%d days, %d hours, %d minutes, %d seconds"), days, hours, minutes, seconds);
+        } else if (hours > 0) {
+            strTime = strprintf(_("%d hours, %d minutes, %d seconds"), hours, minutes, seconds);
+        } else if (minutes > 0) {
+            strTime = strprintf(_("%d minutes, %d seconds"), minutes, seconds);
+        } else {
+            strTime = strprintf(_("%d seconds"), seconds);
+        }
+    }
+    return strTime;
+}
+
+boost::optional<int64_t> SecondsLeftToHeight(const Consensus::Params& params, int currentHeight, int futureHeight)
```

Comment:
The output of this function will be incorrect if `currentHeight` and `futureHeight` are on opposite sides of a network upgrade that changes the target spacing. The current usage avoids this by only ever setting `futureHeight` to the height of the next epoch, and then using `futureHeight - 1` here. Please canonicalize this by refactoring the function into:

```suggestion
boost::optional<int64_t> SecondsLeftToNextEpoch(const Consensus::Params& params, int currentHeight)
```

and calling `NextActivationHeight()` inside this function. (The improved clarity is worth calling `NextActivationHeight()` twice.)